src/mono/mono/metadata/icall-eventpipe.c

```diff
@@ -132,6 +248,249 @@ eventpipe_thread_detach (void)
 		mono_thread_internal_detach (current_thread);
 }
 
+static
+void
+eventpipe_fire_method_events (
+	MonoJitInfo *ji,
+	EventPipeFireMethodEventsData *events_data)
+{
+	g_assert_checked (ji != NULL);
+	g_assert_checked (events_data->domain != NULL);
+	g_assert_checked (events_data->method_events_func != NULL);
+
+	uint64_t method_id = 0;
+	uint64_t module_id = 0;
+	uint64_t method_code_start = (uint64_t)ji->code_start;
+	uint32_t method_code_size = (uint32_t)ji->code_size;
+	uint32_t method_token = 0;
+	uint32_t method_flags = 0;
+	uint8_t kind = MONO_CLASS_DEF;
+	char *method_namespace = NULL;
+	const char *method_name = NULL;
+	char *method_signature = NULL;
+
+	//TODO: Optimize string formatting into functions accepting GString to reduce heap alloc.
+
+	MonoMethod *method = jinfo_get_method (ji);
+	if (method) {
+		method_id = (uint64_t)method;
+		method_token = method->token;
+
+		if (mono_jit_info_get_generic_sharing_context (ji))
+			method_flags |= METHOD_FLAGS_SHARED_GENERIC_METHOD;
+
+		if (method->dynamic)
+			method_flags |= METHOD_FLAGS_DYNAMIC_METHOD;
+
+		if (!ji->from_aot && !ji->from_llvm) {
+			method_flags |= METHOD_FLAGS_JITTED_METHOD;
+			if (method->wrapper_type != MONO_WRAPPER_NONE)
+				method_flags |= METHOD_FLAGS_JITTED_HELPER_METHOD;
+		}
+
+		if (method->is_generic || method->is_inflated)
+			method_flags |= METHOD_FLAGS_GENERIC_METHOD;
+
+		method_name = method->name;
+		method_signature = mono_signature_full_name (method->signature);
+
+		if (method->klass) {
```

Comment:
`method->klass` will be set for wrapper methods too (for some wrapper types it's arbitrarily set to `System.Object`). Is that reasonable?

Reply:
I have seen some wrappers in CoreClr's events as well, so I believe we could start with that and adjust later if we hit issues.